### PR TITLE
PTX-15700 -  Verify  locally attached fastpath volumes after a node reboot.

### PR DESCRIPTION
**Pull request by gpal1**
**What this PR does / why we need it**:

Verifies the locally attached fastpath active volume remains in fastpath active state after a reboot.

Test case steps:
Step 1: Find all the Storage nodes and set fastpath labels on any one of them
Step 2: Schedule application on the selected storage Node
Step 3: Get app volumes and Check fast path is active on the node
Step 4: Reboot the node wait for it to complete
Step 5: Get app volumes and Check fast path is active on the node

Pass Logs:
root@10.13.8.170:/root/git/go/src/github.com/torpedo/PR_PASS_LOGS

**Which issue(s) this PR fixes** (optional)
Closes # PTX-15700 

**Special notes for your reviewer**:
1. fastpath is currently supported only for locally attached repl 1, since adding a new custom app i.e fio-fastpath-repl1 
2. Added a generic vps name i.e fastpath-repl1-vps for both fastpath and vdbench workload

### CODE REVIEW

File: tests/basic/volume_ops_test.go [hunk 1]
```diff
@@ -1466,3 +1469,136 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+			log.Infof("The pxNodes %v ", pxNodes)
+
+			// Select the first node for the test
+			pxNode = pxNodes[0]
+			log.Infof("The Selected node as Fast path label is %v : ", pxNode)
+
+			// Remove if node-type label is set before the test
+			for _, node := range pxNodes {
+				log.Infof("Node Name: %s\n", node.Name)
+				Inst().AppList = applist
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:1505]: why applist is getting assigned in loop?

File: tests/basic/volume_ops_test.go [hunk 2]
```diff
@@ -1466,3 +1469,136 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+		})
+
+		stepLog = "Step 2: Schedule application and Add label on the selected storage Node"
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			var err error
+
+			/* defer for later removal of label node-type
+			defer func() {
+				Inst().AppList = applist
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:1519]: remove if not required

File: tests/basic/volume_ops_test.go [hunk 3]
```diff
@@ -1466,3 +1469,136 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+	stepLog := "Create fastpath Volume Reboot Node and check if fastpath is active"
+	It(stepLog, func() {
+		log.InfoD(stepLog)
+		stepLog = "Step 1: Find all the Storage nodes and and select a node for test"
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			// Get all the Nodes
+			pxNodes, err := GetStorageNodes()
+			log.FailOnError(err, "Unable to get the storage nodes")
+			log.Infof("The pxNodes %v ", pxNodes)
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:1496]: no need to log all the node details

File: tests/basic/volume_ops_test.go [hunk 4]
```diff
@@ -1466,3 +1469,136 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+
+			// Select the first node for the test
+			pxNode = pxNodes[0]
+			log.Infof("The Selected node as Fast path label is %v : ", pxNode)
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:1500]: just log node name

File: tests/basic/volume_ops_test.go [hunk 5]
```diff
@@ -1466,3 +1469,136 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+			/* defer for later removal of label node-type
+			defer func() {
+				Inst().AppList = applist
+				err = Inst().S.RemoveLabelOnNode(pxNode, k8s.NodeType)
+				log.FailOnError(err, "error removing label on node [%s]", pxNode.Name)
+			}()
+			*/
+
+			// Add label on the selected node
+			err = Inst().S.AddLabelOnNode(pxNode, k8s.NodeType, k8s.FastpathNodeType)
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:1526]: remove the label at the end of the test

File: tests/basic/volume_ops_test.go [hunk 6]
```diff
@@ -1466,3 +1469,136 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+	var pxNode node.Node
+	var contexts []*scheduler.Context
+	var applist = Inst().AppList
+	stepLog := "Create fastpath Volume Reboot Node and check if fastpath is active"
+	It(stepLog, func() {
+		log.InfoD(stepLog)
+		stepLog = "Step 1: Find all the Storage nodes and and select a node for test"
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			// Get all the Nodes
```
Comment (vprabhakar-px) [inline tests/basic/volume_ops_test.go:1493]: can you update the comment as Get all storage nodes

File: tests/basic/volume_ops_test.go [hunk 7]
```diff
@@ -1466,3 +1469,136 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+			pxNodes, err := GetStorageNodes()
+			log.FailOnError(err, "Unable to get the storage nodes")
+			log.Infof("The pxNodes %v ", pxNodes)
+
+			// Select the first node for the test
+			pxNode = pxNodes[0]
```
Comment (vprabhakar-px) [inline tests/basic/volume_ops_test.go:1499]: can we take the random node instead of pxNodes[0] ?

File: tests/basic/volume_ops_test.go [hunk 8]
```diff
@@ -1466,3 +1469,136 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+			log.Infof("The Selected node as Fast path label is %v : ", pxNode)
+
+			// Remove if node-type label is set before the test
```
Comment (vprabhakar-px) [inline tests/basic/volume_ops_test.go:1502]: can we move it to common.go ? and pass labels to remove as argument ? so that any label specified can be removed ?

**Review by vprabhakar-px — CHANGES_REQUESTED** [review]

File: tests/basic/volume_ops_test.go (hunk 4 above)
Comment (gpal1) [reply tests/basic/volume_ops_test.go:1500]: Done

File: tests/basic/volume_ops_test.go (hunk 5 above)
Comment (gpal1) [reply tests/basic/volume_ops_test.go:1526]: made changes as suggested

File: tests/basic/volume_ops_test.go [hunk 9]
```diff
@@ -1484,3 +1488,135 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+			if len(pxNodes) > 0 {
+				pxNode = GetRandomNode(pxNodes)
+			} else {
+				log.FailOnError(errors.New("No Storage Node Availiable"), "Error occured while selecting StorageNode")
+			}
+
+			log.Infof("The Selected node for Fast path label is %v : ", pxNode.Name)
+
+			// Remove if node-type label is set before the test
+			Inst().AppList = applist
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:1526]: why is this required here?

File: tests/basic/volume_ops_test.go [hunk 10]
```diff
@@ -1484,3 +1488,135 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+						}
+					}
+
+				})
+			}
+		})
+		stepLog = " Step 4: Reboot the node wait for it to complete"
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			rebootNodeAndWaitForReady(&pxNode)
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:1577]: before rebooting can you make sure the volume is attached to this node only

File: tests/basic/volume_ops_test.go [hunk 11]
```diff
@@ -1484,3 +1488,135 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+
+		stepLog = "Step 2: Schedule application and Add label on the selected storage Node"
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			var err error
+
+			// Add label on the selected node
+			err = Inst().S.AddLabelOnNode(pxNode, k8s.NodeType, k8s.FastpathNodeType)
+			log.FailOnError(err, fmt.Sprintf("Failed add label on node %s", pxNode.Name))
+			Inst().AppList = []string{"fio-fastpath-repl1"}
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:1538]: make sure it is reverted to initial list at the end of the test

File: tests/basic/volume_ops_test.go (hunk 10 above)
Comment (gpal1) [reply tests/basic/volume_ops_test.go:1600]: Fixed. The new check via IsVolumeAttachedOnNode will take care of this case

File: tests/basic/volume_ops_test.go (hunk 11 above)
Comment (gpal1) [reply tests/basic/volume_ops_test.go:1543]: Fixed

File: tests/basic/volume_ops_test.go (hunk 9 above)
Comment (gpal1) [reply tests/basic/volume_ops_test.go:1531]: Cleaned the unused variable

Comment (lsrinivas-pure) [reply tests/basic/volume_ops_test.go:1531]: @gpal1 still the variable exists. can you please check

File: tests/basic/volume_ops_test.go [hunk 12]
```diff
@@ -1484,3 +1488,157 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+					// Get the app volume details
+					log.FailOnError(err, "Failed to get volumes for app %s", ctx.App.Key)
+					dash.VerifyFatal(len(appVolumes) > 0, true, "App volumes exist?")
+					log.Infof("App volumes details: %v ", appVolumes)
+
+					// Store the volume details
+					for _, vol := range appVolumes {
+						apivol, err := Inst().V.InspectVolume(vol.ID)
+						if err != nil {
+							log.FailOnError(err, "failed to inspect volume %v", vol.ID)
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:1571]: if check is not required here

File: tests/basic/volume_ops_test.go [hunk 13]
```diff
@@ -1484,3 +1488,157 @@ var _ = Describe("{ResizeVolumeAfterFull}", func() {
...
+		})
+		stepLog = " Step 4: Reboot the node wait for it to complete"
+		Step(stepLog, func() {
+			var err error
+			var volExists bool
+			log.Infof(" Before reboot check if the volumes are attached on the local node")
+			for _, volumePtr := range volumrlidttr {
+				volExists, err = Inst().V.IsVolumeAttachedOnNode(volumePtr, pxNode)
+				if !volExists || err != nil {
+					log.FailOnError(err, "Volume attached on local node validation failed")
```
Comment (lsrinivas-pure) [inline tests/basic/volume_ops_test.go:1595]: if err is nil, log.FailOnError will not fail

File: tests/common.go [hunk 14]
```diff
@@ -7792,6 +7792,39 @@ func GetPoolCapacityUsed(poolUUID string) (float64, error) {
...
+	randomIndex := rand.Intn(len(pxNodes))
+	randomNode := pxNodes[randomIndex]
+	return randomNode
+}
+
+func RemoveLabelsAllNodes(label string, forStorage, forStorageLess bool) {
+
+	if forStorage {
+		pxNodes, err := GetStorageNodes()
+		if err != nil {
```
Comment (lsrinivas-pure) [inline tests/common.go:7807]: err nil check not required here

File: tests/common.go [hunk 15]
```diff
@@ -7792,6 +7792,39 @@ func GetPoolCapacityUsed(poolUUID string) (float64, error) {
...
+func GetRandomNode(pxNodes []node.Node) node.Node {
+	rand.Seed(time.Now().UnixNano())
+	randomIndex := rand.Intn(len(pxNodes))
+	randomNode := pxNodes[randomIndex]
+	return randomNode
+}
+
+func RemoveLabelsAllNodes(label string, forStorage, forStorageLess bool) {
+
+	if forStorage {
```
Comment (lsrinivas-pure) [inline tests/common.go:7805]: you can traverse all the nodes at once rather than having separate loops for storage and storageless nodes

Comment (gpal1) [reply tests/common.go]: Optimised this code to avoid an extra loop , we will now traverse the element  of PxNodes only added a check to get both all nodes i.e storage and storageless node via GetStorageDriverNodes()

File: tests/basic/volume_ops_test.go (hunk 13 above)
Comment (gpal1) [reply tests/basic/volume_ops_test.go]: Thanks for pointing that FailOnError already does  `if err != nil ` check. fixed it

**Review by vprabhakar-px — APPROVED** [review]
